### PR TITLE
BUG: Initialize the full nditer buffer in case of error

### LINKED ISSUE #18810: Segfault when summing array slice from Pandas dataframe
A segmentation fault is generated when converting a Pandas dataframe to a NumPy array, with incompatible types, and trying to sum a slice.

I have bisected the commit that generated the regression: https://github.com/numpy/numpy/commit/b946f0995a183273df5eec7efbb6c2a74bc058f5 (#18469, #18450) - cc @ahaldane @seberg 

### Reproducing code example:

Requirements:
```
numpy==1.20.2
pandas==1.2.4
```

Code:
```python
import datetime

import numpy as np
import pandas

df = pandas.DataFrame(data={
    'x': [datetime.datetime.now() for _ in range(3)],
    'y': [0 for _ in range(3)],
})
np.sum(df.to_numpy()[:2])
```

### Error message:

GDB traceback:

```
$ gdb --args python test.py
GNU gdb (GDB) 10.1

For help, type "help".
Type "apropos word" to search for commands related to "word"...
Reading symbols from python...
(No debugging symbols found in python)
(gdb) run
Starting program: /home/mike/.virtualenvs/numpy/bin/python test.py
[Thread debugging using libthread_db enabled]
Using host libthread_db library "/usr/lib/libthread_db.so.1".
[New Thread 0x7ffff4236640 (LWP 10386)]
[New Thread 0x7ffff3a35640 (LWP 10387)]
[New Thread 0x7fffef234640 (LWP 10388)]
[New Thread 0x7fffeea33640 (LWP 10389)]
[New Thread 0x7fffea232640 (LWP 10390)]
[New Thread 0x7fffe7a31640 (LWP 10391)]
[New Thread 0x7fffe5230640 (LWP 10392)]
[New Thread 0x7fffe2a2f640 (LWP 10393)]
[New Thread 0x7fffe022e640 (LWP 10394)]
[New Thread 0x7fffdda2d640 (LWP 10395)]
[New Thread 0x7fffdb22c640 (LWP 10396)]

Thread 1 "python" received signal SIGSEGV, Segmentation fault.
0x00007ffff6b925a8 in PyArray_Item_XDECREF () from /home/mike/.virtualenvs/numpy/lib/python3.9/site-packages/numpy/core/_multiarray_umath.cpython-39-x86_64-linux-gnu.so
(gdb) bt
#0  0x00007ffff6b925a8 in PyArray_Item_XDECREF () from /home/mike/.virtualenvs/numpy/lib/python3.9/site-packages/numpy/core/_multiarray_umath.cpython-39-x86_64-linux-gnu.so
#1  0x00007ffff6b82c85 in npyiter_clear_buffers () from /home/mike/.virtualenvs/numpy/lib/python3.9/site-packages/numpy/core/_multiarray_umath.cpython-39-x86_64-linux-gnu.so
#2  0x00007ffff6b8529e in NpyIter_Deallocate () from /home/mike/.virtualenvs/numpy/lib/python3.9/site-packages/numpy/core/_multiarray_umath.cpython-39-x86_64-linux-gnu.so
#3  0x00007ffff6c0e76f in PyUFunc_ReduceWrapper () from /home/mike/.virtualenvs/numpy/lib/python3.9/site-packages/numpy/core/_multiarray_umath.cpython-39-x86_64-linux-gnu.so
#4  0x00007ffff6d1159b in PyUFunc_GenericReduction () from /home/mike/.virtualenvs/numpy/lib/python3.9/site-packages/numpy/core/_multiarray_umath.cpython-39-x86_64-linux-gnu.so
#5  0x00007ffff6d13a90 in ufunc_reduce () from /home/mike/.virtualenvs/numpy/lib/python3.9/site-packages/numpy/core/_multiarray_umath.cpython-39-x86_64-linux-gnu.so
#6  0x00007ffff7d1fed3 in ?? () from /usr/lib/libpython3.9.so.1.0
#7  0x00007ffff7d1ef0b in PyObject_Call () from /usr/lib/libpython3.9.so.1.0
#8  0x00007ffff7d02cc7 in _PyEval_EvalFrameDefault () from /usr/lib/libpython3.9.so.1.0
#9  0x00007ffff7cfbfdd in ?? () from /usr/lib/libpython3.9.so.1.0
#10 0x00007ffff7d0e01e in _PyFunction_Vectorcall () from /usr/lib/libpython3.9.so.1.0
#11 0x00007ffff7cfe29f in _PyEval_EvalFrameDefault () from /usr/lib/libpython3.9.so.1.0
#12 0x00007ffff7cfbfdd in ?? () from /usr/lib/libpython3.9.so.1.0
#13 0x00007ffff7d0e01e in _PyFunction_Vectorcall () from /usr/lib/libpython3.9.so.1.0
#14 0x00007ffff6ab1af4 in array_implement_array_function () from /home/mike/.virtualenvs/numpy/lib/python3.9/site-packages/numpy/core/_multiarray_umath.cpython-39-x86_64-linux-gnu.so
#15 0x00007ffff7d1ff22 in ?? () from /usr/lib/libpython3.9.so.1.0
#16 0x00007ffff7d063aa in _PyObject_MakeTpCall () from /usr/lib/libpython3.9.so.1.0
#17 0x00007ffff7d01c65 in _PyEval_EvalFrameDefault () from /usr/lib/libpython3.9.so.1.0
#18 0x00007ffff7cfbfdd in ?? () from /usr/lib/libpython3.9.so.1.0
#19 0x00007ffff7d0e01e in _PyFunction_Vectorcall () from /usr/lib/libpython3.9.so.1.0
#20 0x00007ffff7d01acf in _PyEval_EvalFrameDefault () from /usr/lib/libpython3.9.so.1.0
#21 0x00007ffff7cfbfdd in ?? () from /usr/lib/libpython3.9.so.1.0
#22 0x00007ffff7cfb9a1 in _PyEval_EvalCodeWithName () from /usr/lib/libpython3.9.so.1.0
#23 0x00007ffff7dc0913 in PyEval_EvalCode () from /usr/lib/libpython3.9.so.1.0
#24 0x00007ffff7dd0b0d in ?? () from /usr/lib/libpython3.9.so.1.0
#25 0x00007ffff7dcc3fb in ?? () from /usr/lib/libpython3.9.so.1.0
#26 0x00007ffff7c6b695 in ?? () from /usr/lib/libpython3.9.so.1.0
#27 0x00007ffff7c6aa49 in PyRun_SimpleFileExFlags () from /usr/lib/libpython3.9.so.1.0
#28 0x00007ffff7de253a in Py_RunMain () from /usr/lib/libpython3.9.so.1.0
#29 0x00007ffff7db3939 in Py_BytesMain () from /usr/lib/libpython3.9.so.1.0
#30 0x00007ffff7a2db25 in __libc_start_main () from /usr/lib/libc.so.6
#31 0x000055555555504e in _start ()
```

It's relevant to note that the sum should fail, and that it does as expected, when the sum is done for a single row (`np.sum(df.to_numpy()[:1])`), and all the rows using slicing (`np.sum(df.to_numpy()[:3]`) or not (`np.sum(df.to_numpy())`).

The segmentation fault is only present when the array slicing takes a subset of the array rows, greater than 1.

### NumPy/Python version information:

* NumPy: 1.20.2
* Python: 3.9.3
* Within virtualenv (no Docker)
```
>>> import sys, numpy; print(numpy.__version__, sys.version)
1.20.2 3.9.3 (default, Apr  8 2021, 23:35:02)
[GCC 10.2.0]
```

### PR DESCRIPTION
**Pull request by seberg**
This is necessary because in some rare cases (reductions), we may
not actually use the full buffer.  In that case, the cleanup-on-error
code would have to grow smart enough to handle these cases.
It seems much simpler to just always initialize the full buffers, even
if we may not end up using them.

Admittedly, the old logic might have skipped the buffer clearing
(especially the full buffer) in a fair bit of cases, but since
this is only relevant for `object` dtype, I assume this is fine.

---

The second commit removes the current `memset` calls that I am very sure don't make much sense.  But to err on the safe side, we could keep them in for the 1.20.x branch.

Closes gh-18810

### CODE REVIEW

File: numpy/core/tests/test_nditer.py [hunk 1]
```diff
@@ -2920,6 +2920,14 @@ def __bool__(self):
     assert_raises(TypeError, np.logical_or.reduce, 
                              np.array([T(), T()], dtype='O'))
 
+def test_object_iter_cleanup_reduce():
+    # Similar as above, but a complex reduction case that was previously
+    # missed (see gh-18810)/
+    # the following array is special in that it cananot be flattened:
```
Comment (tylerjereddy) [inline numpy/core/tests/test_nditer.py:2926]: cananot -> cannot

File: numpy/core/src/multiarray/nditer_api.c [hunk 2]
```diff
@@ -2529,15 +2532,6 @@ npyiter_copy_to_buffers(NpyIter *iter, char **prev_dataptrs)
                 skip_transfer = 1;
             }
 
-            /* If the data type requires zero-inititialization */
-            if (PyDataType_FLAGCHK(dtypes[iop], NPY_NEEDS_INIT)) {
-                NPY_IT_DBG_PRINT("Iterator: Buffer requires init, "
-                                    "memsetting to 0\n");
-                memset(ptrs[iop], 0, dtypes[iop]->elsize*op_transfersize);
-                /* Can't skip the transfer in this case */
-                skip_transfer = 0;
```
Comment (mattip) [inline numpy/core/src/multiarray/nditer_api.c:2538]: Should the code to reset `skip_transfer = 0` remain? Is there ever a case where both `reuse_reduce_loops && prev_dataptrs[iop] == ad_ptrs[iop]` and `PyDataType_FLAGCHK(dtypes[iop], NPY_NEEDS_INIT)`

Comment (seberg) [reply numpy/core/src/multiarray/nditer_api.c:2538]: The case you are saying seems very plausible to me. Note that `NPY_NEEDS_INIT` is pretty much equivalent to "object dtype", so it isn't much of a limitation.

Since this is limited to object-dtype and for all other dtypes the transfer is skipped, I would be surprised if it was necessary.  But, I have to dig deeper :/...   In the case of an object read-write loop, we have to be sure that the buffer is not cleared to be able to reuse it (this translates to "move references" not being used to write it back to the array).  I guess I have to try and make sure we have a test for this (we may well have, but object dtype tests tend to be a bit spotty).

Comment (seberg) [reply numpy/core/src/multiarray/nditer_api.c:2538]: Well, I think you are right, we have to copy always if:
```
                    transferinfo[iop].write.func != NULL &&
                    PyDataType_REFCHK(transferinfo[iop].write.descriptors[0])
```
(or similar).  But I am still struggling with a reproducer, maybe I have to set up a weirdly complicated nditer manually...

Comment (seberg) [reply numpy/core/src/multiarray/nditer_api.c:2538]: I guess its also a bit confusing that `NEEDS_INIT` and `HASREF` are really synonymous...  Anything that has reference must require initialization.  But I can't see of a reason why the other way around can make sense (requiring initialization but for a reason other than `HASREF`).

Comment (mattip) [reply numpy/core/src/multiarray/nditer_api.c:2538]: The easiest thing to do here would be to remove only line 2536, maybe with a `/* TODO: do we really need this? */`

Comment (seberg) [reply numpy/core/src/multiarray/nditer_api.c:2538]: Yeah, that was just the, not really ideal, try to rephrase the current logic into something slightly more to the point.  I prefer to rephrase it (although by just making it `if (!skip_transfer || PyDataType_REFCHK(dtypes[iop])) {`)

I finally found a reproducer for at least one way this can fail with reductions...

File: numpy/core/tests/test_nditer.py [hunk 3]
```diff
@@ -2920,6 +2920,25 @@ def __bool__(self):
...
+        np.ones((8000, 4, 2), dtype=object)[:, ::2, :],
+        np.ones((8000, 4, 2), dtype=object, order="F")[:, ::2, :],
+        np.ones((8000, 4, 2), dtype=object)[:, ::2, :].copy("F")])
+def test_object_iter_cleanup_large_reduce(arr):
+    # More complicated calls are possible for large arrays:
+    out = np.ones(8000, dtype=np.intp)
+    # force casting with `dtype=object`
+    res = np.sum(arr, axis=(1, 2), dtype=object, out=out)
+    assert_array_equal(res, np.full(8000, 4, dtype=object))
+
```
Comment (mattip) [inline numpy/core/tests/test_nditer.py:2941]: I see. That is quite the edge case.

**Comment by mattip** [comment]
Thanks @seberg
